stdlib/LinearAlgebra/src/generic.jl

```diff
@@ -1416,6 +1416,49 @@ function axpby!(α, x::AbstractArray, β, y::AbstractArray)
     y
 end
 
+"""
+    rot!(x, y, c, s)
+
+Overwrite `x` with `c*x + s*y` and `y` with `-conj(s)*x + c*y`.
+Returns `x` and `y`.
+
+!!! compat "Julia 1.5"
+    `rot!` requires at least Julia 1.5.
+"""
+function rot!(x::AbstractVector, y::AbstractVector, c, s)
+    n = length(x)
+    if n != length(y)
+        throw(DimensionMismatch("x has length $(length(x)), but y has length $(length(y))"))
+    end
+    @inbounds for i = 1:n
+        xi, yi = x[i], y[i]
+        x[i] =       c *xi + s*yi
+        y[i] = -conj(s)*xi + c*yi
+    end
+    return x, y
+end
+
+"""
+    ref!(x, y, c, s)
+
+Overwrite `x` with `c*x + s*y` and `y` with `conj(s)*x - c*y`.
+Returns `x` and `y`.
+
+!!! compat "Julia 1.5"
+    `rot!` requires at least Julia 1.5.
+"""
+function ref!(x::AbstractVector, y::AbstractVector, c, s)
+    n = length(x)
+    if n != length(y)
+        throw(DimensionMismatch("x has length $(length(x)), but y has length $(length(y))"))
+    end
+    @inbounds for i = 1:n
```

Comment:
If you start your indices with 1 here, then you will need to `require_one_based_indexing(x, y)` before, and similarly in `rot!`.

Reply:
Late to the party but I'm just curious: why not allow more general arrays and use `eachindex` here? The documentation of `eachindex` says

> If you supply more than one AbstractArray argument, eachindex will create an iterable object that is fast for all arguments

so it seems possible to iterate over `x` and `y` even though they might have different indexing styles?!

Reply:
I'm not sure, but I believe that `eachindex` throws an error if you pass, say, two vectors of equal length, but one has an offset.